### ISSUE TITLE
[Bug] `on_close` can not be triggered

### ISSUE THREAD

**Issue by msyyc** [websocket-client/websocket-client#899]
When we run the following code:
```python
import websocket
from threading import Thread
import sys
import time

def on_close(ws, close_status_code, close_msg):
    print("### closed ###")

def on_message(ws, message):
    print(message)
    time.sleep(2)

if __name__ == "__main__":
    websocket.enableTrace(True)
    if len(sys.argv) < 2:
        host = "ws://echo.websocket.events/"
    else:
        host = sys.argv[1]
    ws = websocket.WebSocketApp(host,
                                on_message=on_message,
                                on_close=on_close)
    Thread(target=ws.run_forever).start()
    time.sleep(1)
    ws.close()
```
`on_close` can not be triggered. But if we close the ws with this way:
```python
...
    # ws.close()
    ws.sock.close()
```
`on_close` can be called successfully

**Comment by engn33r** [comment]
I am able to recreate the issue using the code you shared. If I remove the `time.sleep(2)` in the on_message callback, the problem is gone. I will have to investigate later why this sleep in on_message causes a problem.

Another idea is to test the code with rel as suggested in [the README example](https://github.com/websocket-client/websocket-client#long-lived-connection).

**Comment by brostosjoined** [comment]
Am experiencing the same problem here. When i use `ws.close()` to close the websocket **from a class**   i doesnt work and brings up this error
`      ws.close(1000)
  TypeError: WebSocketApp.close() takes 1 positional argument but 2 were given`

But the `ws.sock.close()` works

**Comment by engn33r** [comment]
@brostosjoined it is possible the bug and error message you get is different than the original issue. Could you share the [full debug error trace](https://websocket-client.readthedocs.io/en/latest/examples.html#debug-and-logging-options) and the code that is triggering the error?

**Comment by brostosjoined** [comment]
nvm i tried closing the socket with a specific close status code. Its fine

**Comment by engn33r** [comment]
@msyyc I tested the example code you shared again and `on_close` is called if I make this change suggested by @wormsik in #612:

```diff
def on_message(ws, message):
    print(message)
-    time.sleep(2)
+    timer = threading.Timer(2, ws.close)
+    timer.start()
```

Figuring out why `time.sleep()` causes this issue but timer does not requires more debugging time than I have today, but maybe this is one step closer to solving this issue.

**Comment by engn33r** [comment]
This specific example of the `on_close` issue is fixed by #918 so using v1.6.0 should solve this.
